File: 2023/haskell/day02a/Main.hs
module Main ( main ) where

import Data.List
import System.Environment
import System.Exit
import System.IO

totalRed :: Int
totalRed = 12

totalGreen :: Int
totalGreen = 13

totalBlue :: Int
totalBlue = 14

usage :: IO a
usage = do
  progname <- getProgName
  putStrLn $ "usage: " ++ progname ++ " <file>"
  exitWith (ExitFailure 1)

process :: String -> Int
process contents = sum validGames
  where
    validGames = [gameNum | line <- lines contents
                           , let gameStr =  head (splitOn ": " line)
                           , let revealsStr = head $ drop 1 (splitOn ": " line)
                           , let gameNumStr = head $ drop 1 (words gameStr)
                           , let gameNum = read gameNumStr :: Int
                           , let valid = all (\subsetStr ->
                                              all (\cubesStr ->
                                                     let color = head $ drop 1 (words cubesStr)
                                                         amountStr = head (words cubesStr)
                                                         amount = read amountStr :: Int
                                                     in case color of
                                                          "red"   -> amount <= totalRed
                                                          "green" -> amount <= totalGreen
                                                          "blue"  -> amount <= totalBlue
                                                          _       -> error ("unknown color: '" ++ color ++ "'")
                                                  ) (splitOn ", " subsetStr)
                                          ) (splitOn "; " revealsStr)
                           , valid
                           ]

splitOn :: Eq a => [a] -> [a] -> [[a]]
splitOn _ [] = []
splitOn delimiter list =
  chunk : splitOn delimiter rest
  where
    (chunk, rest) = breakList delimiter list

breakList :: Eq a => [a] -> [a] -> ([a], [a])
breakList delimiter list =
  case stripPrefix delimiter list of
    Just suffix -> ([], suffix)
    Nothing ->
      case list of
        [] -> ([], [])
        x : xs -> let (chunk, rest) = breakList delimiter xs in (x : chunk, rest)

main :: IO ()
main = do
  args <- getArgs
  case args of
    [] -> usage
    [filename] -> do
      contents <- readFile filename
      let result = process contents
      putStrLn $ "result = " ++ show result
    _ -> usage

File: 2023/rust/day02a/src/main.rs
use std::env;
use std::fs;
use std::process;

const TOTAL_RED: u32 = 12;
const TOTAL_GREEN: u32 = 13;
const TOTAL_BLUE: u32 = 14;

fn usage() {
    let progname = env::args().next().unwrap();
    println!("usage: {progname} <file>");
    process::exit(1);
}

fn process(contents: &str) -> u32 {
    let mut result: u32 = 0;
    for line in contents.lines() {
        if let Some((game_str, reveals_str)) = line.split_once(": ") {
            if let Some((_, game_num_str)) = game_str.split_once(' ') {
                let game_num: u32 = game_num_str.parse::<u32>().unwrap();
                let mut valid: bool = true;
                for subset_str in reveals_str.split("; ") {
                    for cubes_str in subset_str.split(", ") {
                        if let Some((amount_str, color)) = cubes_str.split_once(' ') {
                            let amount: u32 = amount_str.parse::<u32>().unwrap();
                            match color {
                                "red" => {
                                    if amount > TOTAL_RED {
                                        valid = false;
                                    }
                                }
                                "green" => {
                                    if amount > TOTAL_GREEN {
                                        valid = false;
                                    }
                                }
                                "blue" => {
                                    if amount > TOTAL_BLUE {
                                        valid = false;
                                    }
                                }
                                _ => panic!("unknown color"),
                            }
                        }
                    }
                }
                if valid {
                    result += game_num;
                }
            }
        }
    }
    result
}

fn main() {
    if env::args().count() < 2 {
        usage();
    }
    let filename = env::args().nth(1).unwrap();
    let contents = fs::read_to_string(filename).expect("read of input file failed");
    let result = process(&contents);
    println!("result = {result}");
}
